File: solutions/fsharp/protein-translation/4/ProteinTranslation.fs
module ProteinTranslation


let toProtein codon =
    match codon with
    | "AUG" -> "Methionine"
    | "UUU"
    | "UUC" -> "Phenylalanine"
    | "UUA"
    | "UUG" -> "Leucine"
    | "UCU"
    | "UCC"
    | "UCA"
    | "UCG" -> "Serine"
    | "UAU"
    | "UAC" -> "Tyrosine"
    | "UGU"
    | "UGC" -> "Cysteine"
    | "UGG" -> "Tryptophan"
    | "UAA"
    | "UAG"
    | "UGA" -> "STOP"
    | _ -> failwith "Invalid Codon!"


let proteins rna =
    rna
    |> Seq.chunkBySize 3
    |> Seq.map (System.String >> toProtein)
    |> Seq.takeWhile ((<>) "STOP")
    |> Seq.toList

File: solutions/csharp/protein-translation/2/ProteinTranslation.cs

using System;
using System.Collections.Generic;

public static class ProteinTranslation
{

    public static string[] Proteins(string strand)
    {
        List<string> result = new List<string>(strand.Length / 3);
        for (int i = 0; i < strand.Length; i += 3)
        {
            string translated = strand.Substring(i, 3) switch
            {
                "AUG" => "Methionine",
                "UGG" => "Tryptophan",
                "UUU" or "UUC" => "Phenylalanine",
                "UUA" or "UUG" => "Leucine",
                "UAU" or "UAC" => "Tyrosine",
                "UGU" or "UGC" => "Cysteine",
                "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
                "UAA" or "UAG" or "UGA" => "STOP",
                _ => throw new Exception("Invalid sequence")
            };
            if (translated == "STOP") break;
            result.Add(translated);
        }
        return result.ToArray();
    }
}